File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    ListNode* removeNodes(ListNode* head) {
        
        vector< pair< ListNode* , int> > st; 
        ListNode* h=head;
        while(h!=NULL ){
            if(!st.empty()){
                
                while( !st.empty() && st.back().second <h->val){
                    st.pop_back();
                }
                
                if(!st.empty()){
                    st.back().first->next=h;
                }
            }
            st.push_back({h,h->val});
            h=h->next;    
        }
        return st[0].first;
    }
};

File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNodes(ListNode head) {
        List<ListNode > st=new ArrayList<>();
        ListNode h=head;
        while( h!=null){
            if(!st.isEmpty()){
                while( !st.isEmpty() && st.get(st.size()-1).val < h.val ){
                    st.remove(st.size()-1);
                }
                if(!st.isEmpty()){
                    st.get(st.size()-1).next=h;
                }
            }
            st.add(h);
            h=h.next;
            
        }
        return st.get(0);
    }
}